### DISCUSSION TITLE
How to add new parameters for Term Object in Taxonomy

### DISCUSSION THREAD

**Discussion by karthik-tarento** [Sunbird-Knowlg/Community#92] [General Discussions]
Hi All,

We are exploring the possibilities to add new meta parameters for Term Object using Taxonomy APIs. 
By using Framework Term Create API - we are able to add new params... and newly added params are available when using Framework Term Read API.. 
But the newly added parameters are not available when we use the Framework Read API. 
What changes are required to add new meta parameters for Term object ?

Version used - https://github.com/project-sunbird/sunbird-learning-platform/tree/release-5.2.0

Regards,
Karthik

**Comment by pallakartheekreddy** [comment]
@karthik-tarento To get newly added or updated terms in framework read API, we need to publish the framework. Please try to publish the framework and check.
Can you please share the term's data what you are trying to create for reference.

@maheshkumargangula @Krishnaj20

**Reply by karthik-tarento** [reply]
@pallakartheekreddy -- I'm adding my own attribute for the term object which is not listed in Term definition object.
Let's say new attribute name and value is ---> "reviewLevel": "1"

The same accepted in the Term create API and available back when we call the Term Read API. After publishing the Framework, doing Framework Read API and it doesn't return the "reviewLevel" attribute in response. 

After debugging the code, found that we need to update Neo4J definition and executed below query in Neo4J.
MATCH (n:domain{IL_FUNC_OBJECT_TYPE:"Term",IL_UNIQUE_ID:"DEFINITION_NODE_Term"}) SET n.fields=["identifier", "name", "code", "description", "category", "status", "translations", "reviewLevel"];

After  executing the above query, published the framework again and framework read returns the newly added attribute "reviewLevel".

What is the proper procedure to add new attribute to Term definition? I see few APIs in TaxonomyControllerV3 and GraphV3Controller -- but not able to figure out the request body / file structure required for the definition update API. 
Please let me know how to add new attribute both "String" type and "Object" type. For example, we would like to add the following in Term definition as properties...

"reviewStatus": "1",
"additionalPropperties": {
 "contextType": "Functional",
 "contextSource": "sourceOrgId"
}

Regards,
Karthik

**Reply by Krishnaj20** [reply]
@pallakartheekreddy @maheshkumargangula - fya please.

**Reply by maheshkumargangula** [reply]
@karthik-tarento - For now, the process which you followed is fine. There is a backlog task to move these configurations to deployment (as input config file) and implement the framework APIs under `taxonomy-service`. After implementing these backlog items, it will be a simple configuration change in the Term schema file.
Cc: @Krishnaj20 @pallakartheekreddy

**Reply by kameshbhr** [reply]
@maheshkumargangula Does this mean, that even after the backlog items are done, adding a new term to a category can happen only through config change of Term schema file and not through an API? It would be better if the taxonomy-service provides an API (add/remove term from a category) for this. And similarly for adding/removing a category to/from a framework. Unless there is an API, creating a UI for taxonomy framework management wouldn't be possible.

**Comment by pallakartheekreddy** [comment]
@karthik-tarento Is your query resolved? If so, pls mark it as answered.

**Reply by karthik-tarento** [reply]
@maheshkumargangula -- is there any ticket / trackers -- which I can monitor the same to get the information once the backlog tasks done ??
@pallakartheekreddy -- there is no option to mark this thread as answered.

**Comment by maheshkumargangula** [comment]
@karthik-tarento – Below is the ticket for you to check. Please let us know if you are interested in contributing for the same. 
https://project-sunbird.atlassian.net/browse/KN-10

Cc: @pallakartheekreddy @Krishnaj20

**Reply by pallakartheekreddy** [reply]
@karthik-tarento As we haven't received a response from you, we will proceed to close this thread. If you have any further questions or require assistance in the future, please don't hesitate to start a new thread. Thank you.
